c/xudt_rce.c

```diff
@@ -0,0 +1,481 @@
+consider successful.
+ */
+typedef int (*ValidateFuncType)(int is_owner_mode, size_t extension_index,
+                                const uint8_t* args, size_t args_len);
+
+typedef enum XUDTFlags {
+  XUDTFlags_Plain = 0,
+  XUDTFlags_InArgs = 1,
+  XUDTFlags_InWitness = 2,
+} XUDTFlags;
+
+typedef enum XUDTValidateFuncCategory {
+  XVFC_Normal = 0,  // normal extension script
+  XVFC_RCE = 1,     // Regulation Compliance Extension
+} XUDTValidateFuncCategory;
+
+uint8_t RCE_HASH[32] = {1};
+
+// functions
+int load_validate_func(const uint8_t* hash, uint8_t hash_type,
+                       ValidateFuncType* func, XUDTValidateFuncCategory* cat) {
+  int err = 0;
+  void* handle = NULL;
+  size_t consumed_size = 0;
+
+  if (memcmp(RCE_HASH, hash, 32) == 0) {
+    *cat = XVFC_RCE;
+    *func = rce_validate;
+    return 0;
+  }
+
+  CHECK2(MAX_CODE_SIZE > g_code_used, ERROR_NOT_ENOUGH_BUFF);
+  err = ckb_dlopen2(hash, hash_type, &g_code_buffer[g_code_used],
+                    MAX_CODE_SIZE - g_code_used, &handle, &consumed_size);
+  CHECK(err);
+  CHECK2(handle != NULL, ERROR_CANT_LOAD_LIB);
+  ASSERT(consumed_size % RISCV_PGSIZE == 0);
+  g_code_used += consumed_size;
+
+  *func = (ValidateFuncType)ckb_dlsym(handle, EXPORTED_FUNC_NAME);
+  CHECK2(*func != NULL, ERROR_CANT_LOAD_LIB);
+
+  *cat = XVFC_Normal;
+  err = 0;
+exit:
+  return err;
+}
+
+int verify_script_vec(uint8_t* ptr, uint32_t size, uint32_t* real_size) {
+  int err = 0;
+
+  CHECK2(size >= MOL_NUM_T_SIZE, ERROR_INVALID_MOL_FORMAT);
+  mol_num_t full_size = mol_unpack_number(ptr);
+  *real_size = full_size;
+  CHECK2(*real_size <= size, ERROR_INVALID_MOL_FORMAT);
+  err = 0;
+exit:
+  return err;
+}
+
+int get_structure(uint32_t index, mol_seg_t* item) {
+  int err = 0;
+  if (!g_witness_inited) {
+    uint64_t witness_len = WITNESS_SIZE;
+    err = ckb_checked_load_witness(g_witness, &witness_len, 0, 0,
+                                   CKB_SOURCE_GROUP_INPUT);
+    CHECK(err);
+    g_witness_len = witness_len;
+    g_witness_inited = true;
+  }
+
+  mol_seg_t seg = {.ptr = g_witness, .size = g_witness_len};
+
+  mol_seg_t output = MolReader_WitnessArgs_get_input_type(&seg);
+  CHECK2(MolReader_Bytes_verify(&output, false) == MOL_OK,
+         ERROR_INVALID_MOL_FORMAT);
+
+  mol_seg_t raw = MolReader_Bytes_raw_bytes(&output);
+  mol_seg_t structures = MolReader_XudtWitnessInput_get_structure(&raw);
+  CHECK2(MolReader_BytesVec_verify(&structures, false) == MOL_OK,
+         ERROR_INVALID_MOL_FORMAT);
+
+  mol_seg_res_t structure = MolReader_BytesVec_get(&structures, index);
+  CHECK(structure.errno);
+  CHECK2(MolReader_Bytes_verify(&structure.seg, false) == MOL_OK,
+         ERROR_INVALID_MOL_FORMAT);
+
+  *item = structure.seg;
+
+  err = 0;
+exit:
+  return err;
+}
+
+// the *var_len may be bigger than real length of raw extension data
+int load_raw_extension_data(uint8_t** var_data, uint32_t* var_len) {
```

Comment:
It might be possible to leverage new partial loading design of molecule-c2 to optimize the API here.

Reply:
The moleculec-c2 is used already: it's a big change. Before using this new lib, it was improved a lot:
- add fuzzer test
- fix security bugs, like overflow, out of bounds
see lastest commits:
https://github.com/XuJiandong/moleculec-c2/commits/master

Hope it can also be added to audit. 

It's already on "ckb-c-stdlib", together with "ckb_smt.h":
https://github.com/XuJiandong/ckb-c-stdlib/tree/add-smt
Will be sent as a PR.